pep-0725.rst

```diff
@@ -0,0 +1,236 @@
+        else:
+            # Type checkers can assume val is an 'int' in this branch
+
+Since the ``TypeGuard`` type (or output type) is a subtype of the first input
+parameter type, a type checker can determine that the only possible type in the
+``else`` is the other type in the ``Union``. In this example, it is safe to
+assume that if ``is_str`` returns false, then type of the ``val`` argument is an
+``int``.
+
+
+Specification
+=============
+
+This PEP requires no new changes to the language. It is merely modifying the
+definition of ``TypeGuard`` for type checkers. Runtimes are already behaving
+in this way.
+
+
+Existing ``TypeGuard`` usage may change though, as described in 
+`Backwards Compatibility`_.
+
+Unsafe Narrowing
+--------------------
+
+There are cases where this further type narrowing is not possible though. Here's
+an example:
+
+.. code-block:: python
+
+    def is_str_list(val: list[int | str]) -> TypeGuard[list[str]]
+        return all(isinstance(x, str) for x in val)
+
+    def func(val: list[int | str]):
+        if is_str_list(val):
+            # Type checker assumes list[str] here
+        else:
+            # Type checker cannot assume list[int] here
+
+Since ``list`` is invariant, it doesn't have any subtypes. ``list[str]`` is not
+a subtype of ``list[str | int]``. This means type checkers cannot narrow the
+type to ``list[int]`` in the false case.
+
+Type checkers should not assume any narrowing in the false case when the
+``TypeGuard`` type argument is not a subtype of the first input parameter type. 
+
+However, narrowing in the true case is still possible. In the example above, the
+type checker can assume the list is a ``list[str]`` if the ``TypeGuard``
+function returns true.
+
+User error
+--------------------------
+
+The new ``else`` case for a ``TypeGuard`` can be setup incorrectly. Here's an
+example:
+
+.. code-block:: python
+
+    def is_positive_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes int here
+        else:
+            # Type checker assumes str here
+
+A type checker will assume for the else case that the value is ``str``. This is
+a change in behavior from :pep:`647` but as that pep stated, there are many ways
+a determined or uninformed developer can subvert type safety.
+
+A better way to handle this example would be something like so:
+
+.. code-block:: python
+
+    PosInt = NewType('PosInt', int)
+
+    def is_positive_int(val: PosInt | int | str) -> TypeGuard[PosInt]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes PosInt here
+        else:
+            # Type checker assumes str | int here
+
+
+Backwards Compatibility
+=======================
+
+For preexisting code this PEP should require no changes.
```

Comment:
Hmm, not sure how to word that at the moment. Existing code should continue to behave the same way it does now, but yes type checkers will have different behavior. Which might cause more problems if people assume type checkers are always right. Guido's example comes to mind: https://github.com/python/typing/issues/996#issuecomment-1002716830

Perhaps this should say something like:

`Type checkers will now be making more assumptions about the results of TypeGuard function. This can potentially cause problems like in this example:

def is_positive_int(a: int|str) -> TypeGuard[int]:
    return isinstance(a, int) and a >= 0

However, there's no practical reason why anybody would assume the negative case in this situation wasn't an int. Users can still create TypeGuards that are invalid, but in the majority case, the TypeGuard will now behave in a more consistent manor.`

Reply:
I'm not sure I follow completely. This PEP's only proposed change is to type checker behavior, so that's the backward compatibility we're concerned about.

Somebody who used a TypeGuard like in Guido's linked example is following the PEP 647 spec. With this PEP's proposed change, type checkers might not work correctly on their codebase any more:

```
def print_it(x: int | str) -> None:
    if is_positive_int(x):
         print("it's positive!")
    else:
         print(f"it's a string: {len(x)}")
```

With PEP 647 semantics, type checkers would catch the mistake in the else block. With this PEP's semantics, they would not.

Now, you can make the argument that such code is contrived and unlikely to appear in practice, as @erictraut did in https://github.com/python/typing/issues/996#issuecomment-1002728309.

But the PEP needs to acknowledge that it *does* break compatibility, and discuss why that is OK and how we can mitigate the possible harm. (For example, could type checkers temporarily raise a warning on encountering a TypeGuard that will now behave differently?)

Reply:
I don't think type checkers should be trying to check for all possible scenarios where this might happen. As in PEP 647, people can make invalid TypeGuards.

In Guido's example, pyright at least couldn't tell that something might happen. It doesn't understand the logic of the TypeGuard. 

I think there is no mitigation other than people that write TypeGuards being smart enough to know when they're doing something that a typechecker can't possibly catch.